src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+using Microsoft.AspNetCore.Rewrite.Logging;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Rewrite.Internal
+{
+    public class RedirectToWwwRule : IRule
+    {
+        public int StatusCode { get; set; }
+        public bool AllowLocalhost { get; set; } = true;
+
+        public virtual void ApplyRule(RewriteContext context)
+        {
+            var req = context.HttpContext.Request;
+            if (AllowLocalhost && req.Host.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
+            {
+                context.Result = RuleResult.ContinueRules;
+                return;
+            }
+
+            if (req.Host.Value.StartsWith("www.", StringComparison.OrdinalIgnoreCase))
+            {
+                context.Result = RuleResult.ContinueRules;
+                return;
+            }
+
+            var wwwHost = new HostString($"www.{req.Host}");
+            var newUrl = UriHelper.BuildAbsolute(req.Scheme, wwwHost, req.PathBase, req.Path, req.QueryString);
```

Comment:
Add in the FragmentString here (req.FragmentString). It was a mistake for other rules to not include it.

Reply:
how would I get the # fragment at this point? We're on the server

Reply:
No, there's never a fragment on the request.